script/src/verify.rs

```diff
@@ -1029,50 +1008,184 @@ impl<DL: CellDataProvider + HeaderProvider + ExtensionProvider + Send + Sync + C
+            vec![ResumableMachine::initial(machine)]
+        };
+
+        run_vms(script_group, max_cycles, machines, &context)
+    }
+}
+
+// Run a series of VMs that are just freshly resumed
+fn run_vms(
+    script_group: &ScriptGroup,
+    max_cycles: Cycle,
+    mut machines: Vec<ResumableMachine>,
+    context: &Arc<Mutex<MachineContext>>,
+) -> Result<ChunkState, ScriptError> {
+    // (exit_code, cycles)
+    let mut callee_data = None;
+    let mut spawn_data = None;
+
+    if machines.is_empty() {
+        return Err(ScriptError::VMInternalError(
+            "To resume VMs, at least one VM must be available!".to_string(),
+        ));
+    }
+
+    let map_vm_internal_error = |error: VMInternalError| match error {
+        VMInternalError::CyclesExceeded => ScriptError::ExceededMaximumCycles(max_cycles),
+        _ => ScriptError::VMInternalError(format!("{error:?}")),
+    };
+
+    while !machines.is_empty() {
+        let mut machine = machines.pop().unwrap();
+
+        if let (Some((callee_exit_code, callee_cycles)), Some(callee_spawn_data)) =
+            (callee_data, &spawn_data)
+        {
+            update_caller_machine(
+                &mut machine.machine_mut().machine,
+                callee_exit_code,
+                callee_cycles,
+                callee_spawn_data,
+            )
+            .map_err(map_vm_internal_error)?;
         }
 
         match machine.run() {
             Ok(code) => {
-                if code == 0 {
-                    Ok(ChunkState::Completed(machine.machine.cycles()))
-                } else {
-                    Err(ScriptError::validation_failure(&script_group.script, code))
+                callee_data = Some((code, machine.cycles()));
+                if let ResumableMachine::Spawn(_, data) = machine {
+                    spawn_data = Some(data);
                 }
             }
             Err(error) => match error {
                 VMInternalError::CyclesExceeded => {
-                    Ok(ChunkState::suspended(ResumableMachine::new(machine, None)))
+                    let mut new_suspended_machines: Vec<_> = {
+                        let mut context = context.lock().map_err(|e| {
+                            ScriptError::VMInternalError(format!("Failed to acquire lock: {}", e))
+                        })?;
+                        context.suspended_machines.drain(..).collect()
+                    };
+                    // The inner most machine lives at the top of the vector,
+                    // reverse the list for natural order.
+                    new_suspended_machines.reverse();
+                    let mut suspended_machines: Vec<_> = machines.drain(..).collect();
+                    suspended_machines.push(machine);
+                    suspended_machines.append(&mut new_suspended_machines);
+                    return Ok(ChunkState::suspended(
+                        suspended_machines,
```

Comment:
I think we can use the owned `machines` vec directly:
```suggestion
                    machines.push(machine);
                    machines.append(&mut new_suspended_machines);
                    return Ok(ChunkState::suspended(
                        machines,
```